contrib/tools/stacks-relay/src/http.rs

```diff
@@ -0,0 +1,54 @@
+use std::{collections::HashMap, io::Read};
+
+#[derive(Debug)]
+pub struct RequestMessage {
+    pub method: String,
+    pub url: String,
+    pub protocol: String,
+    pub headers: HashMap<String, String>,
+}
+
+pub trait RequestMessageEx: Read {
+    fn read_http_request_message(&mut self) -> RequestMessage {
+        let mut read_byte = || {
+            let mut buf = [0; 1];
+            self.read_exact(&mut buf).unwrap();
+            buf[0]
+        };
+        let mut read_line = || {
+            let mut result = String::new();
+            loop {
+                let b = read_byte();
+                if b == 13 {
+                    break;
+                };
+                result.push(b as char);
+            }
+            assert_eq!(read_byte(), 10);
+            result
+        };
```

Comment:
Is there any particular reason for defining these as closures instead of helper methods?